### DISCUSSION TITLE
Expand ESP32 NVS to support partition by label.

### DISCUSSION THREAD

**Discussion by brianredbeard** [micropython/micropython#9928] [Ideas]
In the Espressif documentation for the [non-volatile storage library][nvs-docs] 

> To mitigate potential conflicts in key names between different components, NVS assigns each key-value pair to one of namespaces. Namespace names follow the same rules as key names, i.e., the maximum length is 15 characters. Furthermore, there can be no more than 254 different namespaces in one NVS partition. Namespace name is specified in the [nvs_open()](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/nvs_flash.html#_CPPv48nvs_openPKc15nvs_open_mode_tP12nvs_handle_t) or [nvs_open_from_partition](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/nvs_flash.html#_CPPv423nvs_open_from_partitionPKcPKc15nvs_open_mode_tP12nvs_handle_t) call. This call returns an opaque handle, which is used in subsequent calls to the nvs_get_*, nvs_set_*, and [nvs_commit()](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/nvs_flash.html#_CPPv410nvs_commit12nvs_handle_t) functions. This way, a handle is associated with a namespace, and key names will not collide with same names in other namespaces. Please note that the namespaces with the same name in different NVS partitions are considered as separate namespaces.

I ran into an issue with this today when attempting to use my application specific NVS partition.  The partition is clearly seen via the `esp32` module:

```python
>>> esp32.Partition.find(label='appconfig', type=esp32.Partition.TYPE_DATA)
[<Partition type=1, subtype=2, address=4407296, size=524288, label=appconfig, encrypted=0>]
```

Unfortunately though at present only the default ESP32 non-volatile storage partition may be used due to the use of `nvs_open()` as opposed to `nvs_open_from_partition()`:
https://github.com/micropython/micropython/blob/b41aaaa8a918a6645ebc6bfa4483bd17286f9263/ports/esp32/esp32_nvs.c#L67

It would be great if this was supported!

[nvs-docs]: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/nvs_flash.html#namespaces

**Comment by jimmo** [comment]
If you can write a PR to implement this it seems like a pretty reasonable addition. Having a way to pass an optional esp32.Partition instance to the esp32.NVS constructor seems useful.
